File: 2021/c++/src/day18.cpp
#include "AOC.h"
#include <cassert>

list<pii> build(const string& l){
	list<pii> o;
	int level = 0;
	for(const auto& c: l){
		if(c == '[') level++;
		else if(c==']') level--;
		else if(c!=',') o.emplace_back(c-'0', level);
	}
	return o;
}

bool explode(list<pii>& p){
	for(auto it = p.begin(); it!=prev(p.end());it++){
		auto n = next(it);
		if(it->second > 4 && it->second == n->second){
			if(it != p.begin())
				prev(it)->first+=it->first;
			if(it != prev(p.end()))
				next(next(it))->first+=next(it)->first;
			it = p.erase(it);
			it->first=0;
			it->second--;
			return true;
		}
	}
	return false;
}

bool split(list<pii>& p){
	int t;
	for(auto it = p.begin(); it!=p.end();it++){
		if(it->first >= 10){
			t = it->first>>1;
			it->second++;
			p.insert(it, pii(t, it->second));
			it->first=t+(it->first)%2;
			return true;
		}
	}
	return false;
}
void inline reduce(list<pii>& p){
	while (explode(p) || split(p));
}

list<pii> add(list<pii> a, list<pii> b){
	a.splice(a.end(), b);
	for(auto& [k, v]: a)
		v++;
	reduce(a);
	return a;
}

int magnitude(list<pii> l){
	while(l.begin() != prev(end(l))){
		for(auto it = begin(l); it!=prev(end(l));it++){
			auto n = next(it);
			if(it->second == n->second){
				it->first = 3*it->first + 2*n->first;
				it->second--;
				l.erase(n);
				break;
			}
		}
	}
	return l.begin()->first;
}

chrono::time_point<std::chrono::steady_clock> day18(input_t& inp){
	vector<list<pii>> nums;
	for(auto& l: inp)
		nums.push_back(build(l));
	auto s = nums[0];
	for(int i = 1;i<nums.size();i++){
		s = add(s, nums[i]);
	}
	int p1 = magnitude(s), p2(0);
	for(int i = 0;i<inp.size()-1;i++){
		for(int j = i+1;j<inp.size();j++){
			p2 = max(p2, magnitude(add(nums[i], nums[j])));
			p2 = max(p2, magnitude(add(nums[j], nums[i])));
		}
	}
	auto done = chrono::steady_clock::now();
	cout<<"[P1] "<<p1<<"\n[P2] "<<p2<<endl;
	return done;
}

File: 2021/Python/src/day18.py
from utils.aocUtils import *
from itertools import combinations

def build(inp):
	out = []
	level = 0
	for c in inp:
		if c in '[]':
			level+= 1 if c=='[' else -1
		elif c != ',':
			out.append([int(c), level])
	return out
def split(arr):
	for i, v in enumerate(arr):
		if v[0]>=10:
			arr[i:i+1] = [[v[0]//2, v[1]+1], [v[0]//2+(v[0]%2), v[1]+1]]
			return True
def explode(arr):
	for i, (n, l) in enumerate(arr):
		if l>=5:
			if i>0:
				arr[i-1][0]+=n
			if i+2<len(arr):
				arr[i+2][0]+=arr[i+1][0]
			arr[i:i+2] = [[0, l-1]]
			return True
def update(arr):
	while True:
		if explode(arr):
			continue
		if split(arr):
			continue
		break
def magnitude(t):
	arr = t
	while True:
		for i in range(len(arr)-1):
			if arr[i][1] == arr[i+1][1]:
				arr[i:i+2] = [[3*arr[i][0]+2*arr[i+1][0], arr[i][1]-1]]
				break
		else:
			break
	return arr[0][0]
def add(a, b):
	c = [[n, l+1] for n,l in a+b]
	update(c)
	return c
def main(input:str):
	n = build(input.splitlines()[0])
	for l in input.splitlines()[1:]:
		n = add(n, build(l))
	p1 = magnitude(n)
	p2 = 0
	for a, b in combinations(input.splitlines(), 2):
		p2 = max(p2, magnitude(add(build(a), build(b))), magnitude(add(build(b), build(a))))
	return (p1, p2)